### ISSUE TITLE
Adjust cases each of the navigation observers are notified

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#3649]
* `BeforeLeaveObserver` should be notified once for any component that is attached to the UI before the new navigation state is applied.
  * The main use case is to give anyone a chance to `postpone` the navigation for asking the user whether they want to save unsaved changes.
  * All components that are attached before applying the new navigation state are notified, regardless of whether they will become detached. The reason for this is that a parameter change might also warrant showing a dialog that asks the user whether they want to save pending changes.
  * The event and listener should be clearly javadoced to say that the old navigation state might be left, but that it might still be canceled. It should also be pointed out that leaving a navigation state doesn't necessarily mean that all receiving components will be detached. Components should use e.g. `onDetach` if they're interested in actual detaches.
* `BeforeEnterObserver` should be notified once for any component that would be attached to the UI after the new navigation state is applied (but notifying should happen before actually attaching anything).
  * The main use case is to redirect the user to some other location (potentially an error view) e.g. because of authentication.
  * All components that would be attached after applying the navigation state are notified, regardless of whether they were attached previously as well.
  * Obviously, anything that gets attached indirectly through logic in e.g. `onAttach` or `beforeEnter` cannot be considered here.
  * The event and listener should be clearly javadoced to say that a new navigation state might be entered, but that it might still be canceled. It should also be pointed out that the event is also delivered to components that were attached already before the event. Components should use e.g. `onAttach` if they're interested in actual attaches.
* `AfterNavigationObserver` should be notified once for any component that is actually attached after applying the navigation state.
  * The main use case is to allow updates to e.g. breadcrumbs or the active state of router links. Making any "global" changes (e.g. triggering a new navigation) is not recommended at this point (but it's still not explicitly forbidden).

**Comment by denis-anisimov** [comment]
See https://vaadin.slack.com/archives/C6X43FE8M/p1520342919000182?thread_ts=1520329566.000543&cid=C6X43FE8M about `BeforeEnter`.

Actually, as I understand similar thing should be done for `BeforeLeave` : the components which will be detached after the navigation should be calculated before the navigation happens.
That's not the current behavior.

**Comment by Legioth** [comment]
`BeforeLeave` should be fired to everything that is attached before the navigation happens, regardless of whether it will become detached or not. I'll add a clarification about that in the original description as well.

**Comment by denis-anisimov** [comment]
Too many events :(
Anyway it's good that it will be clarified.

But original intention was `BeforeEnter` event: should it be fired for all attached components so that they are calculated before the navigation happens ?

**Comment by Legioth** [comment]
Yes, my idea for `BeforeEnter` would be to calculate which components will be attached.

I think we can find those by collecting descendants from each of the components in the new `chain`, but stop each individual traversal when encountering anything that is explicitly defined as a member (i.e. not include descendants) of either the new or the old `chain`.

**Comment by denis-anisimov** [comment]
>will be attached
That's OK.

But I'm talking about the component which will be _DETACHED_ to avoid notifying them.
(_WILL BE ATTACHED_ I treat as : _they are detached now and will be attached)
To prevent this issue: https://github.com/vaadin/flow/issues/3614
At the moment it's prevented but using improper way: siblings are not notified.
But as I understand all components with _ATTACHED_ state should be notified.
May be the logic you suggest will allow to do this out of the box. 
So we are talking about the same thing already.

**Comment by Legioth** [comment]
I believe we can avoid notifying anything that will be detached with the algorithm I tried to describe above.

We could treat each item in either the old or new `chain` as a root and collect descendants that are "owned" (i.e. don't traverse other roots) by each root in the new chain. In that way, every component that is "owned" by roots that are only in the old chain would be excluded, and each component that would be attached with the new `chain` would be traversed exactly once.

**Comment by denis-anisimov** [comment]
OK

**Comment by heruan** [comment]
Would it be possible to hook an observer *before* the target component instance is constructed?

The target component might depend on a bean that is not available for some reason (e.g. an authenticated user injected in the constructor) so I need to decide to reroute before trying to instantiate the target.

**Comment by pleku** [comment]
This issue touches slightly the refactoring proposed in https://github.com/vaadin/flow/issues/3648

**Comment by pleku** [comment]
The [documentation](https://vaadin.com/docs/v10/flow/routing/tutorial-routing-lifecycle.html) is/will be incorrect, so it needs to be updated to match actual behavior. It should really clearly describe for every observer that what 
- when it is triggered and what it means
- what operations can be done from the listener in relation to the navigation

**Comment by Legioth** [comment]
> Would it be possible to hook an observer before the target component instance is constructed?

One way of doing that would be https://github.com/vaadin/flow/issues/2754.

Otherwise, we would have to define some way of notifying that wouldn't be based on instances, e.g. something like `@NavigationObserver(SomeObserver.class)` on a `ParentLayout` or `@Route` class. I'd treat that as a separate issue instead of doing it as part of this one.

**Comment by Legioth** [comment]
The implementation has been updated, but should still also check what needs to change in the documentation.

**Comment by Legioth** [comment]
Documentation update in https://github.com/vaadin/flow-and-components-documentation/pull/54

**Comment by Legioth** [comment]
Documentation also merged, so apparently nothing more to do.
